### ISSUE TITLE
Crash in NonblockingRng::GenerateBlock

### ISSUE THREAD

**Issue by janblome** [weidai11/cryptopp#391]
On windows `NonblockingRng::GenerateBlock` initializes a static reference to the `MicrosoftCryptoProvider` (see [osrng.cpp](/weidai11/cryptopp/blob/CRYPTOPP_5_6_5/osrng.cpp#L153)). This can lead to a crash when the code was compiled with Visual Studio 2013 because local static variables are not thread-safe, as mentioned in #372.

**Comment by noloader** [comment]
Thanks @janblome.

It looks like we are going to have to forgo C++11 synchronization gear on Microsoft platforms.

**Comment by noloader** [comment]
OK, so I *think* this is could be a problem with [Dynamic Initialization and Destruction with Concurrency](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2008/n2660.htm). Dynamic Initialization and Destruction with Concurrency is a core language feature set forth in 2008. I can't imagine Microsoft implementations lack it considering its almost a decade old, Clang provided it in 2.8, and GCC provided in in 4.3. Additionally, [CPP Reference | C++ 2011 features ](http://en.cppreference.com/w/cpp/compiler_support) states Microsoft compilers support it since `CL` 14.00/VS 2008.

Here's what I am thinking for `config.h`. It adds a define for `CRYPTOPP_MSC_VERSION` and basically disables it for almost all Microsoft compilers and some platforms. Until we figure out what ***does not*** crash under MS toolchains, programs written for Windows will just use the old C++03 Singleton. The old Singleton could leak memory, but it does not crash and burn.

```
// Dynamic Initialization and Destruction with Concurrency
// MS at VS2018 (20.00); GCC at 4.3; Clang at 2.9; Intel 11.1; SunCC 5.13.
#if (CRYPTOPP_MSC_VERSION >= 2000) && ((WINVER >= 0x0600 /*_WIN32_WINNT_VISTA*/) || (_WIN32_WINNT >= 0x0600 /*_WIN32_WINNT_VISTA*/))
#  define CRYPTOPP_CXX11_DYNAMIC_INIT 1
#elif (CRYPTOPP_LLVM_CLANG_VERSION >= 30300) || (CRYPTOPP_APPLE_CLANG_VERSION >= 50000)
#  define CRYPTOPP_CXX11_DYNAMIC_INIT 1
#elif (__INTEL_COMPILER >= 1110)
#  define CRYPTOPP_CXX11_DYNAMIC_INIT 1
#elif (CRYPTOPP_GCC_VERSION >= 40300)
#  define CRYPTOPP_CXX11_DYNAMIC_INIT 1
#elif (__SUNPRO_CC >= 0x5130)
#  define CRYPTOPP_CXX11_DYNAMIC_INIT 1
#endif // Dynamic Initialization compilers
```

Then we change the Singleton implementation in `misc.h`:

```
template <class T, class F, int instance>
  const T & Singleton<T, F, instance>::Ref(CRYPTOPP_NOINLINE_DOTDOTDOT) const
{
#if defined(CRYPTOPP_CXX11_ATOMICS) && defined(CRYPTOPP_CXX11_SYNCHRONIZATION) && defined(CRYPTOPP_CXX11_DYNAMIC_INIT)
    // new C++11 Singleton
# else
    // old C++03 Singleton
#endif
}
```

**Comment by MarcelRaad** [comment]
@noloader 
> Additionally, CPP Reference | C++ 2011 features states Microsoft compilers support it since CL 14.00/VS 2008

They don't list the cl.exe version, but the Visual Studio version, so it's VS 2015, which is correct.

**Comment by denisbider** [comment]
> They don't list the cl.exe version, but the Visual Studio version, so it's VS 2015, which is correct.

... and this is not the type of support Crypto++ wants, since it *does not* work properly on Windows XP and Windows Server 2003 (platforms deprecated by MS, but supported by Crypto++) in common conditions (use of Crypto++ in a DLL).

Therefore, there is effectively no support for this feature, as far as Crypto++ is concerned, on MS platforms.

**Comment by MarcelRaad** [comment]
@denisbider 
> ... and this is not the type of support Crypto++ wants, since it does not work properly on Windows XP and Windows Server 2003 (platforms deprecated by MS, but supported by Crypto++) in common conditions (use of Crypto++ in a DLL).

Of course, that's bad.

But as there is a check for the target Windows version in the above patch, CRYPTOPP_MSC_VERSION can be changed from 2000 to 1900 (VS 2015).

**Comment by noloader** [comment]
Thanks @MarcelRaad and @denisbider,

> MR: They don't list the cl.exe version, but the Visual Studio version, so it's VS 2015, which is correct.

Thanks. I guess it was wishful thinking. An issue was raised on their wiki talk page at http://en.cppreference.com/w/Talk:cpp/compiler_support#MSVC_version_numbers_are_confusing.

> DB: this is not the type of support Crypto++ wants, since it does not work properly on Windows XP and Windows Server 2003 (platforms deprecated by MS, but supported by Crypto++) in common conditions (use of Crypto++ in a DLL).

I guess this is where I struggle. The original Singleton class was added to Crypto++ 5.2 in March 2003. If the original Singleton worked with Crypto++ 5.2 through 5.6.2 on Windows 2000 through Windows 7/8, then how does it no longer work now?

When I say "it works", it includes the original limitation of leaking memory if two threads try to initialize the singleton. Here's the [original code comments from Crypto++ 5.2](https://github.com/weidai11/cryptopp/blob/40cb58f1f41ae823809f0c8e94097b6c50477113/misc.h):

> This function safely initializes a static object in a multithreaded environment without using locks. It may leak memory when two threads try to initialize the static object at the same time but this should be acceptable since each static object is only initialized once per session.

It looks like I need to ask (please forgive my ignorance): are the code comments wrong?

-----

Here's a reference to the issue DB is referring to. It can be found at [Thread Local Storage (TLS)](https://msdn.microsoft.com/en-us/library/6yh4a9k1.aspx) on MSDN:

> On Windows operating systems before Windows Vista, __declspec( thread ) has some limitations. If a DLL declares any data or object as __declspec( thread ), it can cause a protection fault if dynamically loaded. After the DLL is loaded with LoadLibrary, it causes system failure whenever the code references the __declspec( thread ) data. Because the global variable space for a thread is allocated at run time, the size of this space is based on a calculation of the requirements of the application plus the requirements of all the DLLs that are statically linked. When you use LoadLibrary, you cannot extend this space to allow for the thread local variables declared with __declspec( thread ). Use the TLS APIs, such as TlsAlloc, in your DLL to allocate TLS if the DLL might be loaded with LoadLibrary.

**Comment by denisbider** [comment]
@noloader: As far as I know, the old Singleton implementation is "fine".

I'm just pointing out that no version of Visual Studio actually offers "magic statics" (compiler-provided, thread-safe, static object initialization), which are supposed to be part of C++11. What MS provides is available only from VS 2015 onward (whereas Crypto++ aims to support earlier VS), and even what's provided by VS 2015 doesn't work in DLLs on XP and 2003.

As far as the old Singleton, it's badly implemented. There's no need to allow for the race condition, a lock-less singleton can be implemented with a simple check of a non-volatile flag (1 initialized, 0 not initialized), and an InterlockedIncrement of a separate volatile flag (if InterlockedIncrement returns 1, this thread is responsible for initialization; if it returns something else, another thread is already on it; spin until the other thread completes).

**Comment by noloader** [comment]
@MarcelRaad , @denisbider,

> As far as the old Singleton, it's badly implemented....

Yeah, its amazing its lasted so long. There's a lot to be said about "it just works", even when taking into account how some folks feel about pattern vs anti-pattern and other design philosophies.

I estimated it only had a 10% or 25% chance of survival, but it keeps hanging around. I investigated removing it a while back, but I could not figure out how to provide the same/similar behaviors without it. The `NO_OS_DEPENDENCE` use case was a constant problem. Apple, Linux, Solaris, Unix and Windows modulo C++03 and C++11 makes for a rather small tool chest at times.

> ... a lock-less singleton can be implemented with ...

Well, my feeling is, if there's something better than we should consider using it. If you have some time, would you mind sharing some code?

**Comment by noloader** [comment]
I believe we cleared this issue at [Commit 46c9cc725cf325f2](https://github.com/weidai11/cryptopp/commit/46c9cc725cf325f2990056526a00848a11d37fa8).

The commit added `CRYPTOPP_CXX11_DYNAMIC_INIT` to pickup [Dynamic Initialization and Destruction with Concurrency]() conformance. We use the C++11 version if `CRYPTOPP_CXX11_ATOMICS`, `CRYPTOPP_CXX11_SYNCHRONIZATION` and `CRYPTOPP_CXX11_DYNAMIC_INIT` are defined.

**Comment by noloader** [comment]
I believe we cleared this issue at [Commit 46c9cc725cf325f2](http://github.com/weidai11/cryptopp/commit/46c9cc725cf325f2990056526a00848a11d37fa8).

The commit added `CRYPTOPP_CXX11_DYNAMIC_INIT` to pickup [Dynamic Initialization and Destruction with Concurrency]() conformance. We use the C++11 version if `CRYPTOPP_CXX11_ATOMICS`, `CRYPTOPP_CXX11_SYNCHRONIZATION` and `CRYPTOPP_CXX11_DYNAMIC_INIT` are defined. Microsoft platforms have an additional restriction of Windows Vista or above.

Here are the related bug reports for the commit:

* [Issue 372, Singleton::Ref() crashes on MSVC11/12](http://github.com/weidai11/cryptopp/issues/372)
* [Issue 373, Loader lock on WinXP](http://github.com/weidai11/cryptopp/issues/373)
* [Issue 389, Race condition in InitializeInteger::InitializeInteger](http://github.com/weidai11/cryptopp/issues)
* [Issue 391, Crash in NonblockingRng::GenerateBlock](http://github.com/weidai11/cryptopp/issues/391)

**Comment by noloader** [comment]
I believe we cleared this issue at [Commit 46c9cc725cf325f2](http://github.com/weidai11/cryptopp/commit/46c9cc725cf325f2990056526a00848a11d37fa8).

The commit added `CRYPTOPP_CXX11_DYNAMIC_INIT` to pickup [Dynamic Initialization and Destruction with Concurrency]() conformance. We use the C++11 version if `CRYPTOPP_CXX11_ATOMICS`, `CRYPTOPP_CXX11_SYNCHRONIZATION` and `CRYPTOPP_CXX11_DYNAMIC_INIT` are defined. Microsoft platforms have an additional restriction of Windows Vista or above.

Here are the related bug reports for the commit:

* [Issue 372, Singleton::Ref() crashes on MSVC11/12](http://github.com/weidai11/cryptopp/issues/372)
* [Issue 373, Loader lock on WinXP](http://github.com/weidai11/cryptopp/issues/373)
* [Issue 389, Race condition in InitializeInteger::InitializeInteger](http://github.com/weidai11/cryptopp/issues)
* [Issue 391, Crash in NonblockingRng::GenerateBlock](http://github.com/weidai11/cryptopp/issues/391)

**Comment by noloader** [comment]
I believe we cleared this issue at [Commit 46c9cc725cf325f2](http://github.com/weidai11/cryptopp/commit/46c9cc725cf325f2990056526a00848a11d37fa8).

The commit added `CRYPTOPP_CXX11_DYNAMIC_INIT` to pickup [Dynamic Initialization and Destruction with Concurrency]() conformance. We use the C++11 version if `CRYPTOPP_CXX11_ATOMICS`, `CRYPTOPP_CXX11_SYNCHRONIZATION` and `CRYPTOPP_CXX11_DYNAMIC_INIT` are defined. Microsoft platforms have an additional restriction of Windows Vista or above.

Here are the related bug reports for the commit:

* [Issue 372, Singleton::Ref() crashes on MSVC11/12](http://github.com/weidai11/cryptopp/issues/372)
* [Issue 373, Loader lock on WinXP](http://github.com/weidai11/cryptopp/issues/373)
* [Issue 389, Race condition in InitializeInteger::InitializeInteger](http://github.com/weidai11/cryptopp/issues)
* [Issue 391, Crash in NonblockingRng::GenerateBlock](http://github.com/weidai11/cryptopp/issues/391)

**Comment by noloader** [comment]
I believe we cleared this issue at [Commit 46c9cc725cf325f2](http://github.com/weidai11/cryptopp/commit/46c9cc725cf325f2990056526a00848a11d37fa8).

The commit added `CRYPTOPP_CXX11_DYNAMIC_INIT` to pickup [Dynamic Initialization and Destruction with Concurrency]() conformance. We use the C++11 version if `CRYPTOPP_CXX11_ATOMICS`, `CRYPTOPP_CXX11_SYNCHRONIZATION` and `CRYPTOPP_CXX11_DYNAMIC_INIT` are defined. Microsoft platforms have an additional restriction of Windows Vista or above.

Here are the related bug reports for the commit:

* [Issue 372, Singleton::Ref() crashes on MSVC11/12](http://github.com/weidai11/cryptopp/issues/372)
* [Issue 373, Loader lock on WinXP](http://github.com/weidai11/cryptopp/issues/373)
* [Issue 389, Race condition in InitializeInteger::InitializeInteger](http://github.com/weidai11/cryptopp/issues/389)
* [Issue 391, Crash in NonblockingRng::GenerateBlock](http://github.com/weidai11/cryptopp/issues/391)

**Comment by noloader** [comment]
I believe we cleared this issue at [Commit 46c9cc725cf325f2](http://github.com/weidai11/cryptopp/commit/46c9cc725cf325f2990056526a00848a11d37fa8).

The commit added `CRYPTOPP_CXX11_DYNAMIC_INIT` to pickup [Dynamic Initialization and Destruction with Concurrency]() conformance. We use the C++11 version if `CRYPTOPP_CXX11_ATOMICS`, `CRYPTOPP_CXX11_SYNCHRONIZATION` and `CRYPTOPP_CXX11_DYNAMIC_INIT` are defined. Microsoft platforms have an additional restriction of Windows Vista or above.

Here are the related bug reports for the commit:

* [Issue 372, Singleton::Ref() crashes on MSVC11/12](http://github.com/weidai11/cryptopp/issues/372)
* [Issue 373, Loader lock on WinXP](http://github.com/weidai11/cryptopp/issues/373)
* [Issue 389, Race condition in InitializeInteger::InitializeInteger](http://github.com/weidai11/cryptopp/issues/389)
* [Issue 391, Crash in NonblockingRng::GenerateBlock](http://github.com/weidai11/cryptopp/issues/391)

**Comment by janblome** [comment]
Thanks a lot for the quick fix for the issue with the Singleton, but there is still a problem in `NonblockingRng::GenerateBlock`. I created a pull request with a potential fix to show the problem: #392.

**Comment by denisbider** [comment]
The following is new code (untested) which, as far as I know, should work on all versions of Windows:

    enum { CallOnceThreshold = 0x40000000 };
    unsigned int volatile callOnceState = 0;
    bool called = false;
    
    // ...

    if (!called)
        if (InterlockedExchangeAdd(&callOnceState, 0) < CallOnceThreshold)
        {
            if (InterlockedIncrement(&callOnceState) == 1)
            {
                try
                {
                    callee();
                }
                catch (...)
                {
                    called = true;
                    InterlockedExchange(&callOnceState, CallOnceThreshold);
                    throw;
                }

                called = true;
                InterlockedExchange(&callOnceState, CallOnceThreshold);
            }
            else
            {
                while (InterlockedExchangeAdd(&callOnceState, 0) < CallOnceThreshold)
                    SwitchToThread();
            }
        }

In a cross-platform library like Crypto++, which intends to support Visual Studio versions before 2015, and which intends to support XP and Windows Server 2003 as target platforms, I would suggest using this for Windows. For other platforms, can use something that works there.
